WordPress/Classes/ViewRelated/Menus/MenusViewController.m

```diff
@@ -0,0 +1,945 @@
+        [numberScanner scanUpToCharactersFromSet:characterSet intoString:NULL];
+        [numberScanner scanCharactersFromSet:characterSet intoString:&nameNumberStr];
+        
+        if ([nameNumberStr integerValue] > highestInteger) {
+            highestInteger = [nameNumberStr integerValue];
+        }
+    }
+    highestInteger = highestInteger + 1;
+    NSString *menuStr = NSLocalizedString(@"Menu", @"The default text used for filling the name of a menu when creating it.");
+    return [NSString stringWithFormat:@"%@ %i", menuStr, highestInteger];
+}
+
+- (MenuItemEditingViewController *)editingControllerWithItem:(MenuItem *)item
+{
+    MenuItemEditingViewController *controller = [[MenuItemEditingViewController alloc] initWithItem:item blog:self.blog];
+    void(^dismiss)() = ^() {
+        [self dismissViewControllerAnimated:YES completion:nil];
+    };
+    controller.onSelectedToSave = ^() {
+        [WPAppAnalytics track:WPAnalyticsStatMenusEditedItem withBlog:self.blog];
+        [self setNeedsSave:YES forMenu:self.selectedMenuLocation.menu significantChanges:YES];
+        [self.itemsViewController refreshViewWithItem:item focus:YES];
+        dismiss();
+    };
+    controller.onSelectedToTrash = ^() {
+        [WPAppAnalytics track:WPAnalyticsStatMenusDeletedItem withBlog:self.blog];
+        if (item.itemID.integerValue || item.menu.menuID.integerValue == MenuDefaultID) {
+            // If the item had an ID, saving is enabled.
+            // Or if the user trying to edit the default menu, saving is enabled.
+            // Otherwise the item was never created remotely, so no need to save this deletion.
+            [self setNeedsSave:YES forMenu:self.selectedMenuLocation.menu significantChanges:YES];
+        }
+        [self.itemsViewController removeItem:item];
+        [self insertBlankMenuItemIfNeeded];
+        dismiss();
+    };
+    controller.onSelectedToCancel = dismiss;
+    return controller;
+}
+
+#pragma mark - Bar button items
+
+- (void)discardChangesBarButtonItemPressed:(id)sender
+{
+    [WPAppAnalytics track:WPAnalyticsStatMenusDiscardedChanges withBlog:self.blog];
+    [self promptForDiscardingChangesByTheLeftBarButtonItem:^{
+        [self discardAllChanges];
+    } cancellation:nil];
+}
+
+- (void)saveBarButtonItemPressed:(id)sender
+{
+    [self.detailsViewController resignFirstResponder];
+    
+    // Buckle up, we gotta save this Menu!
+    Menu *menuToSave = self.updatedMenuForSaving ?: self.selectedMenuLocation.menu;
+    
+    BOOL defaultMenuEnabled = [self defaultMenuEnabledForSelectedLocation];
+    
+    // Check if user is trying to save the Default Menu and made changes to it.
+    if (menuToSave.menuID.integerValue == MenuDefaultID && defaultMenuEnabled && self.hasMadeSignificantMenuChanges) {
+        
+        // Create a new menu to use instead of the Default Menu.
+        Menu *newMenu = [Menu newMenu:self.blog.managedObjectContext];
```

Comment:
The logic for deriving a new menu from the default menu might have a better home in the menu service vs the controller.